src/gather.jl

```diff
@@ -0,0 +1,82 @@
+export gather, gather!
+
+"""
+    gather!(dst, src, idx)
+
+Reverse operation of scatter, which gather data in `src` to destination according to `idx`.
+For each index `k` in `idx`, assign values to `dst` according to
+
+    dst[:,...,k...] = src[:, ..., idx[k]...]
+
+# Arguments
+- `dst`: the destination where data would be assigned to.
+- `src`: the source data to be assigned.
+- `idx`: the mapping for assignment from source to destination.
+"""
+function gather!(dst::AbstractArray{Tdst,Ndst}, 
+                 src::AbstractArray{Tsrc,Nsrc}, 
+                 idx::AbstractArray{NTuple{M,Int}, Nidx}) where {Tdst,Tsrc,Ndst,Nsrc,Nidx,M}
+  # TODO: use M = _length(eltype(idx)) to merge the integer method into this?
+  @boundscheck _gather_checkbounds(src, idx)
+  Ndst - Nidx == Nsrc - M  || throw(ArgumentError(""))
+  size(dst)[1:Ndst-Nidx] ==  size(src)[1:Ndst-Nidx] || throw(ArgumentError(""))
+  size(dst)[Ndst-Nidx+1:end] ==  size(idx) || throw(ArgumentError(""))
+  
+  coldst = ntuple(i -> Colon(), Ndst - Nidx)
+  colsrc = ntuple(i -> Colon(), Nsrc - M)
+  @simd for k in CartesianIndices(idx)
+    @inbounds view(dst, coldst..., Tuple(k)...) .= view(src, colsrc..., idx[k]...)
```

Comment:
We usually don't use performance annotations in FluxML

Reply:
we do now

Reply:
There's typically no difference in ml routines, so we still don't

Reply:
just to give some context, this is a port (+generalization) of code from https://github.com/yuehhua/ScatterNNlib.jl
that has been around for a while, it's battle-tested, perfomant, and has been benchmarked against similar frameworks. 
The port will be time-consuming and nontrivial, so please let's not make it more painful than needed and let's not undo other people's work.
Bounds checking outside sensitive loops and going inbounds inside is perfectly reasonable, done all over the place in Base, and it's here for good reason.
So the answer to "We usually don't use performance annotations in FluxML" is "So what?" (and a secondary answer is that's the reason we are slow and using loopvectorization or similars has been discussed and will happen at some point).
Please make focused and clear comments, like
"we shouldn't be doing *a specific thing* because *a specific issue* could arise in a *specific situation*" 
and keep in mind the context I just mentioned, otherwise, this is going to be a nightmare.

Reply:
For performance annotations, @DhairyaLGandhi may want to check it [here](https://github.com/FluxML/NNlib.jl/pull/255#issuecomment-787475179). I have done a small benchmark on `scatter!` w/o annotations.

Reply:
Cool, seems like it's not doing much in terms of uptick. Let's leave it out. For folks who need it, they can add those to the function call anyway, and doing it here leaves us open to cases where custom arrays don't check bounds appropriately (or array of structs which may not simd cleanly) for one reason or another. Re vectorisation etc: we would rather consult experts who can give us an idea of better memory accesses etc and that work is ongoing, so let's not jump the gun.

Reply:
I tested without annotations as well in this pr and the perfomance impact is negligible, so I remove them